lighthouse-core/audits/third-party-summary.js

```diff
@@ -85,18 +99,27 @@ class ThirdPartySummary extends Audit {
    * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
    * @param {Array<LH.Artifacts.TaskNode>} mainThreadTasks
    * @param {number} cpuMultiplier
-   * @return {{byEntity: Map<ThirdPartyEntity, Summary>, byURL: Map<string, Summary>, urls: Map<ThirdPartyEntity, string[]>}}
+   * @return {SummaryMap}
    */
   static getSummaries(networkRecords, mainThreadTasks, cpuMultiplier) {
     /** @type {Map<string, Summary>} */
     const byURL = new Map();
-    /** @type {Map<ThirdPartyEntity, Summary>} */
+    /** @type {Map<ThirdPartyEntity, EntitySummary>} */
     const byEntity = new Map();
     const defaultSummary = {mainThreadTime: 0, blockingTime: 0, transferSize: 0};
 
     for (const request of networkRecords) {
-      const urlSummary = byURL.get(request.url) || {...defaultSummary};
+      const urlSummary = byURL.get(request.url) ||
+        {...defaultSummary, firstStartTime: Infinity, firstEndTime: Infinity};
       urlSummary.transferSize += request.transferSize;
+      urlSummary.firstStartTime = Math.min(urlSummary.firstStartTime, request.startTime);
+      if (request.timing) {
+        // Lookahead preparser may start requests before entire HTML is loaded
```

Comment:
I'm not really sure how this comment applies here either. Feels like we should just attach the earliest network request object for the URL at this point rather than store these other-audit-specific summary attributes?

Reply:
In theory could we have two requests to the same URL but the request which starts second but ends first? If we store a single request we cannot capture this situation.

Reply:
> In theory could we have two requests to the same URL but the request which starts second but ends first?

What happens in the audit if that situation occurred and we used the earlier one?

Reply:
```
P1 =====================================
P2       ========     R1 ========
```

It could mean a slightly different cutoff time for resources fetched by the product. The request which ends first (`P2`) might be used to start fetching secondary resources (`R1`) before the other request (`P1`) ends. If we use the request which ends second (`P1`), all these secondary resources (`R1`) would not be counted.

This scenario could be avoided if we store based on first end time, but that leads to a different problem:
```
P1 =================
         R1 ===========================
                              R2 ============
```
These two requests to the same resource (`R1`, `R2`) would be counted even though they shouldn't.

Both situations seem pretty unlikely (maybe impossible?) but to avoid both we can't take the start time and end time from the same request.

Reply:
Ah thanks for the explanation :) I see my confusion now. I thought we should just be showing all the assets from the entity under the product anyway, so I was never really concerned with this problem. At least as far as I could tell, the product request in third-party-web identified was never *guaranteed* to be the *first* request related to that product anyway, it was just *definitely* a request for that product. 

Was this your understanding too or was there more work than I realized that the product labelled request is always guaranteed to the be first request related to this product?

(A key example that comes to mind is the Google Analytics loader which tells you to load it through GTM even if that's the only thing you're loading. In that case, none of our proposed handling methods will ever correctly attribute that Google Analytics is the only reason you're requesting this third-party.)

>

Reply:
The product regex in TPW is used to match the FIRST request for a specific product, but any request fetched after this initial product request that belongs to the same entity is also included in the report. The URL for this initial product request is the URL the developer sees when they add one of these third party products (e.g. the source url in the iframe for a YouTube embed).

Reply:
> The product regex in TPW is used to match the FIRST request for a specific product, but any request fetched after this initial product request that belongs to the same entity is also included in the report.

I think we're talking about different things right now :) Taking a step back...

I'm asking as maintainer of third-party-web if I'm supposed to be upholding a contract that all regexes that appear under a `product` identifier must only identify the first request that is ever issued for that product, which wasn't my understanding when we added the product identification there. I thought it was just marking any URL that definitely identified usage of that product on the page. 

If my understanding *was* correct, then I'm hesitant to use that signal here as the end-all-be-all of when the product-related activity started even if it works out in the YouTube case (because also just using the entity works out for the most part too 😉 ). Does that make sense?

If my understanding was wrong, we might want to consider moving this rather Lighthouse-specific "first product request" identification logic over here into Lighthouse rather than in the generic TPW dataset and sorry for leading you astray!

Reply:
I think I see what you're saying now. Your understanding is different from mine, the regexes I put in TPW were meant to be the first request as well as the identifying request.

Moving the first request logic to LH is fine with me. I would still argue that the identifying resource can be used as the starting point for resources that will always be deferrable. However, if there is only one product under its entity then some other entity resources could be deferrable.

What do you think about having exhaustive product lists under entities? This would allow us to recommend deferring the entire entity if there is a single deferrable product vs deferring a single product under that entity.

Reply:
@adamraine @patrickhulce it's hard to tell where this conversation ended up :)

Reply:
> What do you think about having exhaustive product lists under entities? 

If I'm understanding your proposal correctly of ensuring all URLs of each third-party are assigned to a product, seems like an impossible amount of work to commit to. Domain ownership is easier to track and slower moving but tracking slight internal URL changes and dependencies would need an army of volunteers to maintain third-party-web. Finding *one* request that definitely detects a product was already a big step on its own 😅

> it's hard to tell where this conversation ended up

I'm not quite sure either, tbh. I think we're mostly splitting hairs about something that won't have *too* much impact on the results. 


Summary of my position:

1. I'd like to keep the identification logic in third-party-web inline with that project's mission, to identify domains/URLs that definitely belong to a particular entity or product and not let Lighthouse-specific sequencing-based concerns leak into contributions that can or can't be accepted there.
2. I want to surface the *complete* impact of a third-party product on a page here in Lighthouse.
3. I believe just surfacing all of an entity's resources once a product is detected is the simplest way to accomplish two and has minimal risk of overstating the product's impact. 

I think @adamraine and I are on the same page for 1 and 2, and maybe we just disagree about 3?

Reply:
That pretty much sums it up. Patrick is right that splitting resources based on timing won't have too much impact, but it seems like a pure positive aside from the code organization problem that was the original issue in this thread. That being said, I have not found a site where the impact would be significantly different so I think I will get rid of the timings.